source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRButton.mm

```diff
@@ -21,9 +25,58 @@ + (UIButton* )acr_renderButton:(UIView *)view
                                                alpha:1.0]];
 
     CGSize contentSize = [button.titleLabel intrinsicContentSize];
-    button.frame = CGRectMake(0, 0, contentSize.width, contentSize.height);
-    button.contentEdgeInsets = UIEdgeInsetsMake(5,5,5,5);
-
+    [button setFrame:CGRectMake(0, 0, contentSize.width, contentSize.height)];
+    [button setContentEdgeInsets:UIEdgeInsetsMake(5,5,5,5)];
+    
+    std::shared_ptr<AdaptiveCards::BaseActionElement> action = [acoAction element];
+    
+    if( [iconUrl length] != 0 ){
+        NSMutableDictionary *actionsViewMap = [rootView getActionsMap];
+        __block UIImage *img = nil;
+        // Generate key for ImageViewMap
+        NSString *key = [NSString stringWithCString:action->GetId().c_str() encoding:[NSString defaultCStringEncoding]];
+        // Syncronize access to imageViewMap
+        dispatch_sync([rootView getSerialQueue], ^{
+            // if image is available, get it, otherwise cache UIButton, so it can be used once images are ready
+            if(actionsViewMap[key] && [actionsViewMap[key] isKindOfClass:[UIImage class]]) {
+                img = actionsViewMap[key];
+            }
+            else {
+                actionsViewMap[key] = button;
+            }
+        });
+        
+        if(img){
```

Comment:
it would really nice if you can figure out way to store imageView instead of img, so we can reduce the code duplication and shift more work from main UI thread.